tree-3.py

```diff
@@ -0,0 +1,154 @@
+# Definition for a binary tree node.
+# class TreeNode:
+#     def __init__(self, val=0, left=None, right=None):
+#         self.val = val
+#         self.left = left
+#         self.right = right
+#Problem 1: Path Sum II
+#Time Complexity: O(n^2) create copy of elements each time
+#Space Complexity: O(n^2) copy of elements each time
+#Did it work on leetcode:
+'''
+Brute force: test each path from the top, save in an array, if it adds to sum append it to array
+'''
+
+'''
+Main idea:
+Similar to previous problem, we the present array in the call stack , add
+the next value, when we reach the leaf, we sum all the elements in array and check it it's valid.
+We could just keep the sum, and if it's a valid branch go up until the root and save it, although it would save
+space in the callstack, seems too complicated for not much added.
+'''
+class Solution:
+    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
+        res=[]
+        #careful, we have to pass in copies and not the array itself
+        #or else any change in any recursion will affect it
+        def helper(root,elements=[],s=0):
+            #edge case
+            if not root:
+                return
+
+            elements.append(root.val)
+            s+=root.val
+            if not root.left and not root.right:
+                if s==sum:
+                    res.append(elements)
+                return
+            helper(root.left,elements[:],s)
+            helper(root.right,elements[:],s)
+        helper(root)
+        return res
+
+#Problem 1: Path Sum II
+#Time Complexity: O(n)
```

Comment:
For every leaf node (N/2 total leaf nodes), we copy elements into result arraylist O(N) time. So, in total, time complexity would be O(N^2) here

Reply:
I wrongly copy pasted my solutions. These lines reffer to the second solution of Path II. Should be fixed now! Thank you for your time :)